docs/software-architecture-and-design.md

```diff
@@ -289,6 +294,57 @@ To summarize, the best practices are as follows:
 2. Avoid using coroutines with resizeable containers. If they must be used, make sure that the coroutines are allocated on the heap.
 3. Pass data to the coroutine on creation as much as possible, avoid using member variables.
 
+# Finite State Machines
+## What Are Finite State Machines?
+A finite state machine (FSM) is a system with a finite number of states with defined transitions and outputs based on the inputs to the system. In particular, we are interested in hierarchical state machines where we can transition between states in terms of _when_ states should transition (_guards_) and _what_ should happen when transitions occur (_actions_), given a specific input (_event_). Hierarchical state machines are state machines that are composed of one or more FSMs, which we call sub-FSMs. The parent FSM can treat a sub-FSM as a state with _guards_ and _actions_ when transitioning to and from the sub-FSM. When the sub-FSM enters a terminal state, the parent FSM is able to automatically transition to another state.
+
+![Finite State Machine Diagram](images/finite_state_machine_diagram.png)
```

Comment:
The image doesn't render for me, it says it's an empty file?

Reply:
it renders here https://github.com/UBC-Thunderbots/Software/blob/69d3534c43d7ed3635135e61e03e32a9b20c9085/docs/software-architecture-and-design.md. I think the PR diff view doesn't show it well